### PR TITLE
Implement square-bracket text object

### PR DESCRIPTION
**Pull request by sectioneight**
References #287

### CODE REVIEW

File: test/mode/modeNormal.test.ts [hunk 1]
```diff
@@ -268,6 +268,14 @@ suite("Mode Normal", () => {
     });
 
     newTest({
+      title: "Can handle 'ci[' spanning multiple lines",
```
Comment (rebornix) [inline test/mode/modeNormal.test.ts]: consider adding cases for `a[`, `a]`, `i]`?

Comment (sectioneight) [inline test/mode/modeNormal.test.ts]: Yes, I wasn't sure about the level of testing desired, given that the underlying functionality (character matching) is already covered. Happy to add more tests (and that's generally my default), but wanted to be sensitive to test run times if that's a concern for this project.

Comment (rebornix) [inline test/mode/modeNormal.test.ts]: I agree, the underlying functionality is coverd and no idea to test it again. I'll suggest adding tests for corner cases then we can make sure that we don't break this feature if we want to modify the architecture somehow one day.

For example, per [Vim doc](http://vimdoc.sourceforge.net/htmldoc/motion.html#i[) for `a[`, `When used in Visual mode it is made characterwise.` while this case is linewise. Another case is `i"`, we need take escaped `"` into consideration.

Not sure did I cover this kind of corner cases perfectly when I implement `aw/iw`, but we can together push it to the limit, what do you think? I'm not saying your test cases really miss some corner cases(I'm just not sure).

**Comment by rebornix** [comment]
LGTM

**Comment by johnfn** [comment]
Thanks, @sectioneight!
